Config: display user-friendly message when invalid generator is passed

rodrigoprimo opened the pull request:
# Description

This PR improves how `Config::processLongArgument()` handles the `--generator` parameter to display a user-friendly message if an invalid generator name is passed. Before, an invalid generator name caused a fatal error.

I have checked, and it is not possible to add custom generators using third-party code (at least not without adding files to the PHPCS directory). This is because the `Runner` class only loads generators found in the `src/Generator/` directory:

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/3d14d004265ac6e38f3e161316cf51b7b4812414/src/Runner.php#L97-L98

So, it should be fine to limit the list of valid generators added in this PR to only the three generators provided by PHPCS.

## Suggested changelog entry
Config: display a user-friendly message if an invalid generator name is passed in the command line

## Related issues/external references

Fixes #709 


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [x] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

rodrigoprimo commented:
@jrfnl, while working on this PR, I noticed that, besides what I reported in #709, it is possible to trigger a different PHP fatal error if `Generator` is passed as the name of the generator:

```bash
$ phpcs test.php --generator=Generator --standard=Generic
PHP Fatal error:  Uncaught Error: Cannot instantiate abstract class PHP_CodeSniffer\Generators\Generator in src/Runner.php:98
Stack trace:
#0 bin/phpcs(14): PHP_CodeSniffer\Runner->runPHPCS()
#1 {main}
  thrown in src/Runner.php on line 98
```

This happens because PHPCS tries to instantiate anything that it finds in the `src/Generator` directory:

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/3d14d004265ac6e38f3e161316cf51b7b4812414/src/Runner.php#L97-L98

The change suggested in this PR prevents this fatal error from happening, at least when using PHPCS in the command line. But I'm mentioning it anyway in case we want to do something else about it.

rodrigoprimo commented:
Looking into the unrelated test that failed in the CI build.

jrfnl commented:
> Looking into the unrelated test that failed in the CI build.

Eh.. the test failure is not unrelated...

jrfnl commented on tests/Core/Ruleset/ShowSniffDeprecationsTest.php:
```diff
@@ -153,7 +153,7 @@ public static function dataDeprecatedSniffsListDoesNotShowNeedsCsMode()
             'Standard using deprecated sniffs; documentation is requested' => [
                 'standard'       => __DIR__.'/ShowSniffDeprecationsTest.xml',
-                'additionalArgs' => ['--generator=text'],
+                'additionalArgs' => ['--generator=Text'],
```
Please undo this change. This is perfectly valid. Your PR needs a fix, not this unrelated test.

rodrigoprimo replied:
> Eh.. the test failure is not unrelated...

Indeed, I chose a poor word to describe what I meant.

> Please undo this change. This is perfectly valid. Your PR needs a fix, not this unrelated test.

Just checking, did you check the commit message? Do you disagree with my assessment describing why I updated the test?

jrfnl replied:
I did not, but have now and I don't agree with your assessment.

The basic premise for the CLI arguments is that they should be handled case-insensitively, so `text` is perfectly valid and should be accepted.

rodrigoprimo replied:
But on `master`, `phpcs --standard=Generic --generator=text` does not work and throws a fatal error. Or maybe it throws a fatal error on Linux (case-sensitive OS) but not on Windows (case-insensitive OS)?

If that is the case, I failed to consider case-insensitive operating systems in this PR. I'm not sure what would be the best approach for this PR to accommodate this.

Maybe perform a case-insensitive check for the generator name when running on Windows and a case-sensitive check when running on other systems? What I don't like about this approach is that PHPCS would continue to behave differently depending on the OS.

Another option is to change PHPCS to ensure that it behaves consistently across different operating systems and handles the `--generator` CLI argument case-insensitively. 

> The basic premise for the CLI arguments is that they should be handled case-insensitively

I was not aware of that and I can think of one other example of a CLI argument that is not case-insensitive when PHPCS is running on Linux: `--standard=generic`. I'm not sure if there are other cases.

```
phpcs --generator=Text --standard=generic                                              
ERROR: the "generic" coding standard is not installed. The installed coding standards are MySource, PEAR, PSR1, PSR2, PSR12, Squiz, Zend and MyStandard
```

jrfnl replied:
> But on `master`, `phpcs --standard=Generic --generator=text` does not work and throws a fatal error. Or maybe it throws a fatal error on Linux (case-sensitive OS) but not on Windows (case-insensitive OS)?

Exactly. It is currently (unintentionally) OS-dependent, but shouldn't be.

> Another option is to change PHPCS to ensure that it behaves consistently across different operating systems and handles the `--generator` CLI argument case-insensitively.

IMO, that would be the correct thing to do.

> > The basic premise for the CLI arguments is that they should be handled case-insensitively
> 
> I was not aware of that and I can think of one other example of a CLI argument that is not case-insensitive when PHPCS is running on Linux: `--standard=generic`. I'm not sure if there are other cases.

That's outside the scope of this PR. Might be worth opening an issue to investigate.

jrfnl replied:
Re: CLI and case-insensitive: https://github.com/squizlabs/PHP_CodeSniffer/pull/1656#issuecomment-331772466

rodrigoprimo replied:
Thanks for the additional information. I will drop the commit that changed `testDeprecatedSniffsListDoesNotShowNeedsCsMode()` and update this PR to ensure that the error displayed in `Config::processLongArgument()` for invalid generator names treats the value of the `--generator` argument case-insensitively.

> > Another option is to change PHPCS to ensure that it behaves consistently across different operating systems and handles the --generator CLI argument case-insensitively.
>
> IMO, that would be the correct thing to do.

To implement this, I believe it will be necessary to change `Runner::runPHPCS()` to load the correct generator class no matter what is the case of the passed generator argument. This is a separate change not directly related to #709. Do you prefer to review this change in this PR or a separate PR?

jrfnl replied:
Why do you think the case-(in)sensitively should be handled in the `Runner` class instead of in the `Config` class ?

rodrigoprimo replied:
Because I did not consider normalizing the input in the `Config` class when setting `Config::generator`, and was thinking about doing that in the `Runner` class.

To illustrate what I had in mind, assuming `--generator=text` is passed, my plan was to do a case-insensitive check in the `Config` class to see if the generator name passed is valid. Then, on the `Runner` class the code normalizes the generator name (i.e, modifies the generator name to match the case of the desired file/class name).

After seeing your question, I realized that it makes sense to normalize the generator name in the `Config` class, and that is probably what you had in mind. Is this correct?

I was also missing that `Runner::runPHPCS()` creates an instance of `Config`. Since I was missing that, I was worried about `runPHPCS()` being called directly and attempting to instantiate a generator without the validation and normalization performed in the `Config` class.

rodrigoprimo commented:
@jrfnl, I dropped the commit that changed `testDeprecatedSniffsListDoesNotShowNeedsCsMode()` and added a new commit to ensure that the code handles the `--generator` argument in a case-insensitive way. This PR is ready for another review.

jrfnl commented on src/Config.php:
```diff
@@ -198,6 +198,23 @@ class Config
+     *
+     * @var array<string, string>
+     */
+    private static $validGenerators = [
```
Is there a reason for making this property `static` ?

rodrigoprimo replied:
I made this property `static` because I see its value as common for all instances of this class and not something that will change depending on the instance of the class. Do you see other reasons not to make it `static`?

Anyway, I had to change it to an object property to be able to manipulate its value when generating the error message (copying the value of the property before manipulating it seemed too much). See https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/771#discussion_r1893267997

jrfnl commented on src/Config.php:
```diff
@@ -198,6 +198,23 @@ class Config
+     * - Keys: lowercase version of the generator name.
+     * - Values: name of the generator PHP class.
+     *
+     * Note: once support for PHP < 5.6 is dropped, this property should be refactored into a class
+     * constant.
```
```suggestion
     * {@internal Once support for PHP < 5.6 is dropped, this property should be refactored into a class
     * constant.}
```

jrfnl commented on src/Config.php:
```diff
@@ -198,6 +198,23 @@ class Config
+    /**
+     * A list of valid generators.
+     *
+     * - Keys: lowercase version of the generator name.
+     * - Values: name of the generator PHP class.
```
Should this be part of the "long description" of the variable or part of the description of the `@var` type ?

rodrigoprimo replied:
Makes sense, thanks for the suggestion.

jrfnl commented on src/Config.php:
```diff
@@ -1233,7 +1250,21 @@ public function processLongArgument($arg, $pos)
+                $lowerCaseGeneratorName = strtolower($generatorName);
+
+                if (isset(self::$validGenerators[$lowerCaseGeneratorName]) === false) {
+                    $validOptions = implode(', ', array_values(self::$validGenerators));
```
Is the `array_values()` really needed here ?

jrfnl replied:
Also wondering if for readability, the last comma should be replaced with " and " ?

rodrigoprimo replied:
> Is the array_values() really needed here ?

Ops, no. It was added by mistake.

> Also wondering if for readability, the last comma should be replaced with " and " ?

Yes, I think it is slightly better, and I used it in the original version. But then, when I moved to store the list of valid generators in a class property and used `implode()` to get the string with the name of the generators, I removed "and", assuming that the text is good enough without it, still correct and the code is slightly simpler.

As you are asking, I'm assuming you prefer the version with the "and," so I went ahead and added it. I'm not super happy with my approach using `array_pop()`, but I couldn't think of anything better. Please let me know if you have another suggestion.

Using `array_pop()` meant that I had to change `Config::$validGenerators` to an object property instead of a class property which is related to another comment you left.

jrfnl replied:
Ah, but this is supposed to be an unchangeable array (to become a class constant later on as discussed in previous comments) and the use of `array_pop()` explicitly changes the array (by reference) and would block the future change to make this a class constant. See: https://3v4l.org/QrM61

While the change of the property value would not currently be problematic (as a second `--generator=...` argument is ignored anyway), it does make the code less stable, so I'd like to see an alternative solution for this.

rodrigoprimo replied:
I completely missed that. Now, I pushed a new commit that creates the string with the list of generators without changing the original array. Thanks for your help yesterday suggesting this approach, @jrfnl.

jrfnl commented on src/Config.php:
```diff
@@ -1233,7 +1250,21 @@ public function processLongArgument($arg, $pos)
+                if (isset(self::$validGenerators[$lowerCaseGeneratorName]) === false) {
+                    $validOptions = implode(', ', array_values(self::$validGenerators));
+                    $error        = sprintf(
+                        'ERROR: "%s" is not a valid generator. Valid options are: %s.'.PHP_EOL.PHP_EOL,
```
Just a suggestion...

```suggestion
                        'ERROR: "%s" is not a valid generator. The following generators are supported: %s.'.PHP_EOL.PHP_EOL,
```

jrfnl commented on tests/Core/Config/GeneratorArgTest.php:
```diff
@@ -23,37 +23,59 @@ final class GeneratorArgTest extends TestCase
      * Ensure that the generator property is set when the parameter is passed a valid value.
      *
-     * @param string $generatorName Generator name.
+     * @param string $argumentValue         Generator name passed in the command line.
```
```suggestion
     * @param string $argumentValue         Generator name passed on the command line.
```

jrfnl commented on tests/Core/Config/GeneratorArgTest.php:
```diff
@@ -23,37 +23,59 @@ final class GeneratorArgTest extends TestCase
-    public static function dataGeneratorNames()
+    public static function dataValidGeneratorNames()
     {
         return [
```
As this array is now a little more involved, I think it would make it more straight-forward to maintain and to read the tests if the array included test case names + argument names.

rodrigoprimo replied:
Done

jrfnl reviewed (changes requested):
@rodrigoprimo Thanks for updating this. I have left some comments inline, mostly nitpicks/dotting of the i's and crossing the t's.

Other than that, this is looking good to go into the 3.12.0 release.

rodrigoprimo commented:
Thanks for your additional comments, @jrfnl. I pushed a new commit addressing all of them. Please take another look when you get a chance.

jrfnl reviewed (approved):
@rodrigoprimo Thank you for making those updates and for working on this. All looks good to me!

I'm going to rebase the PR before merging it as the last push feel in the time slot when Coveralls were having problems, so code coverage is not recorded (and I'd like to double-check it's okay).
I'll merge once the build has passed again.
